plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/FileSystemExchangeSource.java

```diff
@@ -15,58 +15,72 @@
+import static io.airlift.concurrent.MoreFutures.toCompletableFuture;
+import static io.airlift.concurrent.MoreFutures.whenAnyComplete;
+import static java.lang.Math.min;
 import static java.util.Objects.requireNonNull;
 
 public class FileSystemExchangeSource
         implements ExchangeSource
 {
-    private final FileSystemExchangeStorage exchangeStorage;
     @GuardedBy("this")
-    private final Iterator<URI> files;
+    private final Iterator<ExchangeSourceFile> exchangeSourceFileIterator;
     @GuardedBy("this")
-    private final Iterator<Optional<SecretKey>> secretKeys;
-
-    @GuardedBy("this")
-    private SliceInput sliceInput;
+    private final List<ExchangeStorageReader> readers;
     @GuardedBy("this")
     private boolean closed;
 
-    public FileSystemExchangeSource(FileSystemExchangeStorage exchangeStorage, List<URI> files, List<Optional<SecretKey>> secretKeys)
+    public FileSystemExchangeSource(
+            FileSystemExchangeStorage exchangeStorage,
+            List<ExchangeSourceFile> exchangeSourceFiles,
+            int maxPageStorageSize,
+            int exchangeSourceConcurrentReaders)
     {
-        this.exchangeStorage = requireNonNull(exchangeStorage, "exchangeStorage is null");
-        checkArgument(requireNonNull(files, "files is null").size() == requireNonNull(secretKeys, "secretKeys is null").size(),
-                format("number of files (%d) doesn't match number of secretKeys (%d)", files.size(), secretKeys.size()));
-        this.files = ImmutableList.copyOf(files).iterator();
-        this.secretKeys = ImmutableList.copyOf(secretKeys).iterator();
+        requireNonNull(exchangeStorage, "exchangeStorage is null");
+        this.exchangeSourceFileIterator = ImmutableList.copyOf(requireNonNull(exchangeSourceFiles, "exchangeSourceFiles is null")).iterator();
+
+        int numReaders = min(exchangeSourceFiles.size(), exchangeSourceConcurrentReaders);
+        this.readers = new ArrayList<>(numReaders);
+        for (int i = 0; i < numReaders; ++i) {
+            try {
+                readers.add(exchangeStorage.createExchangeStorageReader(exchangeSourceFileIterator, maxPageStorageSize));
+            }
+            catch (IOException e) {
+                throw new UncheckedIOException(e);
+            }
+        }
     }
 
     @Override
-    public CompletableFuture<Void> isBlocked()
+    public synchronized CompletableFuture<Void> isBlocked()
     {
-        return NOT_BLOCKED;
+        if (readers.isEmpty()) {
+            return NOT_BLOCKED;
+        }
+        return toCompletableFuture(whenAnyComplete(
+                readers.stream()
+                        .map(ExchangeStorageReader::isBlocked)
+                        .collect(toImmutableList())));
     }
 
     @Override
     public synchronized boolean isFinished()
     {
-        return closed || (!files.hasNext() && sliceInput == null);
+        return closed || readers.isEmpty();
```

Comment:
Maybe just make the `close` variable `volatile` and set it to `true` when all readers are finished. Then this method doesn't have to be synchronized.